crates/bevy_ui/src/ui_node.rs

```diff
@@ -34,91 +35,134 @@ pub enum Val {
+            _ => Err(ValArithmeticError::NonIdenticalVariants),
         }
     }
-}
 
-impl MulAssign<f32> for Val {
-    fn mul_assign(&mut self, rhs: f32) {
-        match self {
-            Val::Undefined | Val::Auto => {}
-            Val::Px(value) | Val::Percent(value) => *value *= rhs,
+    pub fn try_add_to_self(&mut self, rhs: Val) -> Result<(), ValArithmeticError> {
+        *self = self.try_add(rhs)?;
+        Ok(())
+    }
+
+    /// Tries to subtract the values of two [`Val`]s.
+    /// Returns [`ValArithmeticError::NonIdenticalVariants`] if two [`Val`]s are of different variants.
+    /// When adding non-numeric [`Val`]s, it returns the value unchanged.
+    pub fn try_sub(&self, rhs: Val) -> Result<Val, ValArithmeticError> {
+        match (self, rhs) {
+            (Val::Undefined, Val::Undefined) => {
+                warn!("Subtracting from Val::Undefined is a redundant operation");
+                Ok(*self)
+            }
+            (Val::Auto, Val::Auto) => {
+                warn!("Subtracting from Val::Auto is a redundant operation");
+                Ok(*self)
+            }
+            (Val::Px(value), Val::Px(rhs_value)) => Ok(Val::Px(value - rhs_value)),
+            (Val::Percent(value), Val::Percent(rhs_value)) => Ok(Val::Percent(value - rhs_value)),
+            _ => Err(ValArithmeticError::NonIdenticalVariants),
         }
     }
-}
 
-impl Div<f32> for Val {
-    type Output = Val;
+    pub fn try_sub_from_self(&mut self, rhs: Val) -> Result<(), ValArithmeticError> {
+        *self = self.try_sub(rhs)?;
+        Ok(())
+    }
 
-    fn div(self, rhs: f32) -> Self::Output {
+    /// A convenience function for simple evaluation of [`Val::Percent`] variant into a concrete [`Val::Px`] value.
+    /// Returns a [`ValArithmeticError::NonEvaluateable`] if the [`Val`] is impossible to evaluate into [`Val::Px`].
+    /// Otherwise it returns a [`Val::Px`] containing the evaluated value.
+    ///
+    /// **Note:** If a [`Val::Px`] value is evaluated, it's returned unchanged.
+    pub fn evaluate(&self, size: f32) -> Result<Val, ValArithmeticError> {
         match self {
-            Val::Undefined => Val::Undefined,
-            Val::Auto => Val::Auto,
-            Val::Px(value) => Val::Px(value / rhs),
-            Val::Percent(value) => Val::Percent(value / rhs),
+            Val::Percent(value) => Ok(Val::Px(size * value / 100.0)),
+            Val::Px(_) => Ok(*self),
+            _ => Err(ValArithmeticError::NonEvaluateable),
         }
     }
-}
 
-impl DivAssign<f32> for Val {
-    fn div_assign(&mut self, rhs: f32) {
-        match self {
-            Val::Undefined | Val::Auto => {}
-            Val::Px(value) | Val::Percent(value) => *value /= rhs,
-        }
+    /// Similar to [`Val::try_add`], but performs [`Val::evaluate`] on both values before adding.
+    /// Both values have to be evaluatable (numeric).
+    pub fn try_add_with_size(&self, rhs: Val, size: f32) -> Result<Val, ValArithmeticError> {
```

Comment:
To be consistent you should add `try_add_assign_with_size` and `try_sub_assign_with_size`.